src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/CodeActions/Razor/ExtractToComponentCodeActionProviderTest.cs

```diff
@@ -0,0 +1,353 @@
+
+            <PageTitle>Home</PageTitle>
+
+            <div id="parent">
+                <[||]div>
+                    <h1>Div a title</h1>
+                    <p>Div a par</p>
+                </div>
+                <div>
+                    <h1>Div b title</h1>
+                    <p>Div b par</p>
+                </div>
+            </div>
+
+            <h1>Hello, world!</h1>
+
+            Welcome to your new app.
+            """;
+        TestFileMarkupParser.GetSpan(contents, out contents, out var selectionSpan);
+
+        var request = new VSCodeActionParams()
+        {
+            TextDocument = new VSTextDocumentIdentifier { Uri = new Uri(documentPath) },
+            Range = VsLspFactory.DefaultRange,
+            Context = new VSInternalCodeActionContext()
+        };
+
+        var context = CreateRazorCodeActionContext(request, selectionSpan, documentPath, contents, supportsFileCreation: true);
+
+        var provider = new ExtractToComponentCodeActionProvider();
+
+        // Act
+        var commandOrCodeActionContainer = await provider.ProvideAsync(context, default);
+
+        // Assert
+        Assert.NotEmpty(commandOrCodeActionContainer);
+    }
+
+    [Fact]
+    public async Task Handle_InProperMarkup_ReturnsEmpty()
+    {
+        // Arrange
+        var documentPath = "c:/Test.razor";
+        var contents = """
+            @page "/"
+            
+            <PageTitle>Home</PageTitle>
+            
+            <div id="parent">
+                <div>
+                    <h1>Div a title</h1>
+                    <p>Div [||]a par</p>
+                </div>
+                <div>
+                    <h1>Div b title</h1>
+                    <p>Div b par</p>
+                </div>
+            </div>
+
+            <h1>Hello, world!</h1>
+
+            Welcome to your new app.
+            """;
+        TestFileMarkupParser.GetSpan(contents, out contents, out var selectionSpan);
+
+        var request = new VSCodeActionParams()
+        {
+            TextDocument = new VSTextDocumentIdentifier { Uri = new Uri(documentPath) },
+            Range = new Range(),
+            Context = new VSInternalCodeActionContext()
+        };
+
+        var context = CreateRazorCodeActionContext(
+            request,
+            selectionSpan,
+            documentPath,
+            contents);
+
+        var provider = new ExtractToComponentCodeActionProvider();
+
+        // Act
+        var commandOrCodeActionContainer = await provider.ProvideAsync(context, default);
+
+        // Assert
+        Assert.Empty(commandOrCodeActionContainer);
+    }
+
+    [Fact]
+    public Task Handle_MultiPointSelection_ReturnsNotEmpty()
+       => TestSelectionStartAndCursorAsync("""
+            @page "/"
+
+            <PageTitle>Home</PageTitle>
+
+            <div id="parent">
+                [|<div>
+                    |]<h1>Div a title</h1>
+                    <p>Div a par</p>
+                </div>$$
```

Comment:
Perhaps it would be clearer to use the named markers for these tests? It's little unclear what this is validating, but I think its this:

```
                {|result:{|selection:<div>
                    |}<h1>Div a title</h1>
                    <p>Div a par</p>
                </div>|}
```

Reply:
I like it!